### ISSUE TITLE
`home.username`, `home.homeDirectory`: "The option is used but not defined"

### ISSUE THREAD

**Issue by piegamesde** [nix-community/home-manager#1548]
### Issue description

I updated home-manager to latest master and rebuilding then failed because `home.username` and `home.homeDirectory` where used but not defined. I don't define them in my home config, but also I don't use them anywhere. 

Specifying those keys in the config makes it build again.

### Meta

I suspect that #1339 broke it, but I didn't bisect anything yet.

#### Maintainer CC

@rycee 

#### Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 5.8.10-arch1-1, Arch Linux, noversion`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels: `"home-manager, nixpkgs-21.03pre244772.502845c3e31"`
 - nixpkgs: `~/.nix-defexpr/channels/nixpkgs`

**Comment by cole-h** [comment]
https://github.com/nix-community/home-manager/blob/8537920706fafab12a2c2a138dfb6f5c35a6b240/modules/home-environment.nix#L183-L202

If your `home.stateVersion` is >= 20.09, `home.username` and `home.homeDirectory` no longer have defaults. You must set them yourself.

**Comment by piegamesde** [comment]
Seen that. But I didn't change anything and it broke through a HM update. Shouldn't `stateVersion` protect me from exactly these kind of breakings?

**Comment by cole-h** [comment]
If your `stateVersion` was 20.09, it wouldn't protect you because 20.09 was not yet officially released, I guess.

I ran into the same problem while I was running from the master branch and had a `stateVersion` of 20.09: https://logs.nix.samueldr.com/home-manager/2020-05-27#1590617615-1590617837.

Basically: if your `stateVersion` is 20.09, you need to specify `home.username` and `home.homeDirectory`. Since h-m only recently branched off to `release-20.09`, stability was not yet guaranteed.

**Comment by piegamesde** [comment]
Wow, thank you. I'm on unstable and I upgraded to `20.09` quite some time ago because of some feature I don't even remember. All error's on me then. Some unordered thoughts on this:

- There should be a news/changelog entry for those who are yet to upgrade
- This makes the `home.username` and `home.homeDirectory` properties effectively required. Worse, they probably are not until some internal code references them. This means that new users will have it working fine, until they set one seemingly random property and suddenly they are required to set these as well. This does not feel good to me.
- The `stateVersion` documentation should make clear which values are stable and which ones are WIP/not released yet.
- Why does the `stateVersion` default to `18.09`? This means that new users will start on a really old state version. (I guess it's for compatibility, but I don't like this.)

**Comment by cole-h** [comment]
Well, there is a file for 20.09's release notes: https://github.com/nix-community/home-manager/blob/master/doc/release-notes/rl-2009.adoc. I'd guess this functions similarly to NixOS release notes in that the user is responsible for reading it to find out incompatibilities between state versions.

`stateVersion` only defaults to 18.09 in case it is unset, because that is around when `stateVersion`s were introduced, if I had to guess (thus, changing the default could break people on older releases' setups). I don't think this should ever be encountered, because the install script sets the `stateVersion` to the current release, anyways: https://github.com/nix-community/home-manager/blob/b5847455066d0191cebd6b140531cc154443e316/home-manager/install.nix#L48

**Comment by piegamesde** [comment]
The whole `stateVersion` thing is a bit unfortunate, but I'll leave it for another issue some other time for now.

Is it possible to make some values required? To assert that they are set? Because doing so for `home.username` and `home.homeDirectory` in the new state version could provide better error information to the user. If instead both properties are intended to remain optional, they should not be referenced by any provided modules IMO.

**Comment by rycee** [comment]
> There should be a news/changelog entry for those who are yet to upgrade

Did you feel that the phrasing in the [release notes](https://rycee.gitlab.io/home-manager/release-notes.html#sec-release-20.09-state-version-changes) was unclear or inadequate? Would you mind outlining how I could improve them? I fear that I'm not the best in writing clearly so I would appreciate pointers in improving in this area, even better feel free to create a pull-request with your suggested improvements.

> This makes the home.username and home.homeDirectory properties effectively required. Worse, they probably are not until some internal code references them. This means that new users will have it working fine, until they set one seemingly random property and suddenly they are required to set these as well. This does not feel good to me.

But this is how all the options in Home Manager/NixOS work, they are not required to be set until they are needed? It is true that these particular options will be necessary in most people's but from what I can tell the error message is pretty clear, e.g., "The option `home.homeDirectory' is used but not defined." In any case it is expected that people read the release notes when they change the state version. I think using the state version of a unreleased version must reasonably be considered to mean that you are willing to deal with unexpected changes, otherwise we would never be able to make this sort of changes.

> The stateVersion documentation should make clear which values are stable and which ones are WIP/not released yet.

The release notes for the development versions, currently [20.09](https://rycee.gitlab.io/home-manager/release-notes.html#sec-release-20.09), state "This is the current unstable branch and the information in this section is therefore not final", which I find adequate since this also covers the state version changes section but if you'd like an improvement in this area it would be definitely welcome, any documentation contributions are happily accepted.

> Why does the stateVersion default to 18.09? This means that new users will start on a really old state version. (I guess it's for compatibility, but I don't like this.)

It is indeed as @cole-h said, the `stateVersion` option was introduced in version 18.09 and we were therefore forced to put that as the default version. Any later Home Manager setup is expected to have an explicit `stateVersion` set.

**Comment by piegamesde** [comment]
> Would you mind outlining how I could improve them?

No, it's all good, this is mainly a RTFM on my side. The only problem is that the information is not at the place someone will stumble upon accidentally when it would be useful. I personally almost only use the options documentation, and the only information that's there on the topic is [`home.stateVersion`](https://rycee.gitlab.io/home-manager/options.html#opt-home.stateVersion). I suggest adding a link to the releases in the `stateVersion` documentation. I'd also document the "recommended default value" because `"18.09"` is not a "sane default".

Note however that my case is a bit an odd one. I don't think I used the install script (don't ask me how I missed that ^^), and I set the `stateVersion` pretty early on to the latest available value. I didn't bother looking at migration paths because there was not really much config to migrate.

> But this is how all the options in Home Manager/NixOS work, they are not required to be set until they are needed? It is true that these particular options will be necessary in most people's but from what I can tell the error message is pretty clear, e.g., "The option `home.homeDirectory' is used but not defined."

The thing is that usually when I get this error, it is *my* code that references undefined configuration. I grepped through my config and did not find any usages, which was confusing. I don't know how common it is for the home-manager provided modules to reference config keys the user should provide, but I'm slightly in favor of making them all required (depending on how many there are and how they are used). But in the end this is not a big deal because the "usual" paths leading to that error (stateVersion upgrade or adding a new config key) are way more obvious to debug than in my case.

> Any later Home Manager setup is expected to have an explicit stateVersion set.

If we can't error out on this one, does home-manager support warnings?
